### ISSUE TITLE
How do you detect if any window/control has focus and should be consuming mouse events or keyboard events?

### ISSUE THREAD

**Issue by kavika13** [Immediate-Mode-UI/Nuklear#155]
I am looking for a UI-agnostic function that will tell me if Nuklear made a change based on the input I just passed into it. Click and drag on a slider or the title bar are both cases of that. Probably need one to detect if keystrokes are consumed as well.

I tried using `nk_item_is_any_active` to detect the mouse case, and it seems to work a lot of the time. However when I click and drag a title bar fast enough, sometimes the function will return `false` and some input will leak through.

I am not quite sure if this is a bug report or a feature request, hence the phrasing of the issue.

**Comment by kalluwa** [comment]
hi @kavika13 , i just met your problem just now
>However when I click and drag a title bar fast enough, sometimes the function will return false and some input will leak through

this is because your mouse pos checking is runing on every frame, when mouse move quick enough, it will always miss the checking,  but actually it just need to check once when mouse down.

here is my solution:
```c++
case SDL_MOUSEBUTTONDOWN:
    if (nk_item_is_any_active(ctx))       //yep, i just found this function from your issue , XD
	curMode = EDIT_MODE;
    break;
case SDL_MOUSEBUTTONUP:
    curMode = GAME_MODE;
    break;
case SDL_MOUSEMOTION:
    if(curMode == GAME_MODE) //only gui is responding if in EDIT_MODE
        motion(evt.motion.x, evt.motion.y); //handle normal 3d movement
    break;
```
hope it will help

**Comment by kavika13** [comment]
Yeah that will work to solve my problem. Thanks for that suggestion.

It does mean I have to keep track of that state in the rest of my application, whereas I wouldn't have had to before. Since the equivalent "control is being dragged" state is somehow already being kept in Nuklear, I figured it would be nice if it could be exposed somehow.

I still am not sure if this `nk_item_is_any_active` API is working how it should or not, or whether a different API could be exposed to handle this.

**Comment by kalluwa** [comment]
what i confirmed(tested) is: nuklear know whether the gui window is being draging or not.

but **nk_item_is_any_active** didn't use this state( name ? emmm... sorry, i don't know ) ,it just check your mouse pos inside the window or not. 

so  i guess that **nk_item_is_any_active** should be named **nk_mouse_is_inside_gui**(not a real function name in nuklear) , and the real  **nk_item_is_any_active** is:
```python
nk_item_is_any_active(xxxxx):
    if  in drag state :
        return true;
    else
        nk_mouse_is_inside_gui(xxxxxx)
```

after that, your **nk_item_is_any_active** in your code should perform normally

**Comment by kalluwa** [comment]
another solution:

step 1:
add code in line 19317:
```c++
/* window movement by dragging */
        left_mouse_down = in->mouse.buttons[NK_BUTTON_LEFT].down;
        left_mouse_clicked = (int)in->mouse.buttons[NK_BUTTON_LEFT].clicked;
        left_mouse_click_in_cursor = nk_input_has_mouse_click_down_in_rect(in,
            NK_BUTTON_LEFT, header, nk_true);
        if (left_mouse_down && left_mouse_click_in_cursor && !left_mouse_clicked) {
            win->bounds.x = win->bounds.x + in->mouse.delta.x;
            win->bounds.y = win->bounds.y + in->mouse.delta.y;
			
			in->mouse.grab = 1; //add this line 

            in->mouse.buttons[NK_BUTTON_LEFT].clicked_pos.x += in->mouse.delta.x;
            in->mouse.buttons[NK_BUTTON_LEFT].clicked_pos.y += in->mouse.delta.y;
            ctx->style.cursor_active = ctx->style.cursors[NK_CURSOR_MOVE];
        }
```
step2:
```
int nk_item_is_any_active(struct nk_context *ctx)
{
	if (ctx->input.mouse.grab) return true; //add this line

    int any_hovered = nk_window_is_any_hovered(ctx);
    int any_active = (ctx->last_widget_state & NK_WIDGET_STATE_MODIFIED);
    return any_hovered || any_active;
}
```
step3:
use **nk_item_is_any_active**  as  you expected
```c++
case SDL_MOUSEBUTTONDOWN:
	break;
case SDL_MOUSEBUTTONUP:
	break;
case SDL_MOUSEMOTION:
	if(!nk_item_is_any_active(ctx))  //finally, this is the real nk_item_is_any_active
		motion(evt.motion.x, evt.motion.y);
	break;
```

btw:
ctx->input.mouse.grab will be clear to 0 in **nk_input_end(ctx)**, so no need to step4

this is risk here:  **ctx->input.mouse.grab**  maybe used for some other function, i use it just as convenience( acctually, it always be 0 when i tested) and your can add your own value to replace it.

**Comment by kalluwa** [comment]
a solution:
there 3 parts need to be modified

step 1:
```c++
struct nk_mouse {
    struct nk_mouse_button buttons[NK_BUTTON_MAX];
    struct nk_vec2 pos;
    struct nk_vec2 prev;
    struct nk_vec2 delta;
    struct nk_vec2 scroll_delta;
    unsigned char grab;
    unsigned char grabbed;
    unsigned char ungrab;
	unsigned char dragingui; // add this line   1st
};

void nk_input_button(struct nk_context *ctx, enum nk_buttons id, int x, int y, int down)
{
    struct nk_mouse_button *btn;
    struct nk_input *in;
    NK_ASSERT(ctx);
    if (!ctx) return;
    in = &ctx->input;
    if (in->mouse.buttons[id].down == down) return;

	//add following two lines  2nd
	if (nk_window_is_any_hovered(ctx)) in->mouse.dragingui = 1;
	else in->mouse.dragingui = 0;

	btn = &in->mouse.buttons[id];
    btn->clicked_pos.x = (float)x;
    btn->clicked_pos.y = (float)y;
    btn->down = down;
    btn->clicked++;
}

int nk_item_is_any_active(struct nk_context *ctx)
{
	if (ctx->input.mouse.dragingui) return true; //add this line         3rd

    int any_hovered = nk_window_is_any_hovered(ctx);
    int any_active = (ctx->last_widget_state & NK_WIDGET_STATE_MODIFIED);
    return any_hovered || any_active;
}
```

step2:
use **nk_item_is_any_active**  as  you expected
```c++
case SDL_MOUSEBUTTONDOWN:
	break;
case SDL_MOUSEBUTTONUP:
	break;
case SDL_MOUSEMOTION:
	if(!nk_item_is_any_active(ctx))  //finally, this is the real nk_item_is_any_active
		motion(evt.motion.x, evt.motion.y);
	break;
```

**Comment by kavika13** [comment]
Hmm. So I just checked and it looks like if I do all the Nuklear window drawing before calling before I call `nk_item_is_any_active` then it never returns `false` while dragging a window.

If I edit `nk_window_is_any_hovered` to check both the current rect and the previous rect, it doesn't seem to matter if I render before or after the call to `nk_item_is_any_active`, it correctly returns `true` for the window dragging scenario. I have no idea what ripple effects that has so that might be a bad idea for other reasons. But that does imply to me that it's checking the "current" window state against the "current" input, and if you haven't drawn the windows yet then they're still going to be in their "previous" state even though you've added new input.


The simplest change to my code is to just to add this before any code that relies on this state:
```
static bool g_is_nuklear_ui_consuming_input = false;
```

Then set that value immediately after I draw my update the entire UI, wherever that is in the update loop:

```
    draw_my_nuklear_ui_(&g_nuklear_context);

    g_is_nuklear_ui_consuming_input = nk_item_is_any_active(&g_nuklear_context) != 0;
```

This doesn't seem to require any changed from the Nuklear lib.

Based on this, I can't really tell if the function should be changed or not. It does seem a bit weird that functions in the API would be out of sync after I feed input in, but before I update the UI.

**Comment by kalluwa** [comment]
1. 
>nk_item_is_any_active then it never returns false while dragging a window

it's weird and my **nk_item_is_any_active** return **true** when you drag an gui window titlebar


2. 
as you said that:
>Then I don't even need to cache that value
one of the readme of nuklear  is
>No global or hidden state
so it's better to not add any global or static state like **g_is_nuklear_ui_consuming_input** into your project, and what i add in **nk_mouse** should be compatible with nuklear, and it maybe the minimal change with your existing code.

**Comment by kalluwa** [comment]
your project might be very complicated and there could be some better solution if dig in nuklear source code.
but it looks good for now,  
![dragtest](https://user-images.githubusercontent.com/9412335/86919763-53c7a900-c15b-11ea-89f1-0a27730bcf61.gif)

**Comment by kavika13** [comment]
As I said in my previous message, I figured out how to get it working. The method I am using doesn't require any changes to Nuklear or to the cross-platform app framework I am using.

There's nothing really complicated about my app. Its functionality is basically exactly like yours is right now. It is going to be a model and animation previewer for a specific game engine.

The difference between my app and your app is that I am using a framework that doesn't let me implement the event loop myself. There is no code in my app that looks like this:
```
while(SDL_PollEvent(&event) {
    // ... <my code goes here>
    switch(event.type) {
    }
}
```
In the framework I am using, it does the while loop for me, and I only get a callback for each event.

The fix I am now using would apply just fine to your app too, without having to change Nuklear. All you have to do is code your app like this:

```
// Use nk_item_is_any_active without any problems at this spot in the code

while(SDL_PollEvent(&event) {
    // call nk_input_* functions here
}
// DO NOT use nk_item_is_any_active here, or it will be out of sync!
// draw ALL your nk UI here

// Use nk_item_is_any_active without any problems at this spot in the code
```
